### PR TITLE
Enable linking multiple Python formulae

### PR DESCRIPTION
**Pull request by carlocab**
- python@3.9: use `altinstall` target
- python@3.10: link into `HOMEBREW_PREFIX`
- python@3.9: cleanup
- python@3.10: cleanup
- Update Python aliases
- versioned_keg_only_allowlist: add Python 3.10

The primary change (see first commit) uses the `altinstall` target in the Python `Makefile` to make sure that Python3.9 only installs files that contain Python's major and minor version. That is, `python@3.9` will now install only `python3.9`, `python3.9-config`, etc., instead of `python3` and `python3-config`.

This will allow users to have both `python@3.9` and `python@3.10` linked into `HOMEBREW_PREFIX` simultaneously. Users who wish to call `python3.9` as `python3` can add `libexec/"bin"` to their `PATH`.

Going forward, I believe this will help simplify migrating to newer Python versions, since there will be no need to make the newer version keg-only while most dependent formulae still depend on the older version of Python.

I've  made a lot of changes here. For ease of review, I placed substantive changes in the first two commits. The third and fourth commits are refactoring, and the ~~last two~~ fifth and sixth update non-formula files.

I have not tested this on Linux. This may require a few more adjustments there.

See discussion at Homebrew/discussions#2776 and documentation at https://github.com/python/cpython#installing-multiple-versions.

Closes #87075.

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

File: Formula/python@3.10.rb
Comment (carlocab) [inline Formula/python@3.10.rb:269]: @XuehaiPan suggested we do
```ruby
def python3
  bin/"python#{version.major_minor}"
end
```
which is a good idea, given the number of times this pattern appears now.

File: Formula/python@3.10.rb
Comment (carlocab) [inline Formula/python@3.10.rb:226]: Can move this into the block below.

File: Formula/python@3.9.rb
Comment (carlocab) [inline Formula/python@3.9.rb:431]:
```suggestion
      Unversioned and major-versioned symlinks `python`, `python3`, `python-config`, `python3-config`, `pip`, `pip3`, etc. pointing to
```

File: Formula/python@3.10.rb
Comment (carlocab) [inline Formula/python@3.10.rb:465]: Should probably keep the original here.

**Comment by carlocab** [comment]
`python@3.9` failed on Linux. Not sure why.

File: Formula/python@3.9.rb
Comment (carlocab) [inline Formula/python@3.9.rb:191]: I should document what `altinstall` does here in a comment.

**Review by Rylan12 — COMMENTED** [review]
Doesn't this mean that if you run `brew install python@3.9` (no alias so no auto upgrade to `python@3.10`) and then have a script that relies on `/opt/homebrew/bin/python3` it will no longer work unless you manually install `python@3.10`?

**Comment by carlocab** [comment]
> Doesn't this mean that if you run `brew install python@3.9` (no alias so no auto upgrade to `python@3.10`) and then have a script that relies on `/opt/homebrew/bin/python3` it will no longer work unless you manually install `python@3.10`?

Yes. But the same thing happens whenever we make `python@3.9` keg-only and link `python@3.10`.

File: Formula/python@3.9.rb
Comment (carlocab) [inline Formula/python@3.9.rb:457]: @Rylan12 this should help a bit with users who really just want `python3` (of whatever version) but did `brew install python@3.9`.

It won't make it go away, since not everyone reads the `caveats`, but it should help a bit.

File: Formula/python@3.10.rb
Comment (carlocab) [inline Formula/python@3.10.rb:54]: I probably also need `"lib/libpython3.so"` here.

**Comment by cho-m** [comment]
How are we handing formulae that depend on `python@3.9`/etc? Specifically, the usage of:
- `"python3"`
- `which("python3")`
- `Formula["python@3.9"].opt_bin/"python3"`

And similarly build scripts that may use `#!/usr/bin/env python3` or other `python3` references.

**Comment by carlocab** [comment]
> How are we handing formulae that depend on `python@3.9`/etc? Specifically, the usage of:
> 
> * `"python3"`
> * `which("python3")`
> * `Formula["python@3.9"].opt_bin/"python3"`
> 
> And similarly build scripts that may use `#!/usr/bin/env python3` or other `python3` references.

Some discussion at https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-1207357179 and the following two comments.

We should probably continue discussion about this particular issue here instead of there to make it easier to follow.

File: Formula/python@3.10.rb
Comment (carlocab) [inline Formula/python@3.10.rb:228]:
```suggestion
      prefix.glob("*.app") { |app| mv app, app.to_s.sub(/\.app$/, " 3.app") }
```

File: Formula/python@3.9.rb
Comment (carlocab) [inline Formula/python@3.9.rb:205]: I wonder if we should even do this anymore. We no longer ship Python 2, nor is there any practical risk of conflict.

On the other hand, changing this might break users who look for this with the ` 3.app` at the end.

**Comment by carlocab** [comment]
Not too many tests broken by this change. Probably formulae that look for `python3.9` as `python3`. Going forward, we should probably avoid `python3` references and just use the major/minor version (or use `python` in `libexec`).

```
Error: 23 failed steps!
brew test --retry --verbose libcython
brew test --retry --verbose libpython-tabulate
brew test --retry --verbose libxml2
brew test --retry --verbose numpy
brew test --retry --verbose pillow
brew test --retry --verbose protobuf
brew test --retry --verbose protobuf@3
brew test --retry --verbose py3cairo
brew test --retry --verbose pybind11
brew test --retry --verbose pygobject3
brew test --retry --verbose pyqt@5
brew test --retry --verbose six
brew test --retry --verbose swift
brew test --retry --verbose cython
brew test --retry --verbose gdal
brew test --retry --verbose gdcm
brew test --retry --verbose gnuradio
brew test --retry --verbose gobject-introspection
brew test --retry --verbose libsigrok
brew test --retry --verbose mapserver
brew test --retry --verbose opencv
brew test --retry --verbose python-tk@3.9
brew test --retry --verbose root
```

**Comment by carlocab** [comment]
- [x] libcython - #107660
- [x] libpython-tabulate - #107662
- [x] libxml2 - #107663
- [x] numpy - #107664
- [ ] pillow - #107665
- [x] protobuf - #107729
- [x] protobuf@3 - #107730
- [x] py3cairo - #107731
- [ ] pybind11
- [ ] pygobject3
- [ ] pyqt@5
- [ ] six
- [ ] swift
- [ ] cython
- [ ] gdal
- [ ] gdcm
- [ ] gnuradio
- [ ] gobject-introspection
- [ ] libsigrok
- [ ] mapserver
- [ ] opencv
- [ ] python-tk@3.9
- [ ] root

**Comment by carlocab** [comment]
`swift` test error is:
```
  'swiftpmtest': error: invalidManifestFormat("<unknown>:0: error: module \'_Builtin_stddef_max_align_t\' is defined in both \'/var/folders/_0/bwq744sj55n9b7yxj53j1rxw0000gq/C/clang/ModuleCache/OVVET4YQAF2U/_Builtin_stddef_max_align_t-2S6J4W9HHDZWV.pcm\' and \'/var/folders/_0/bwq744sj55n9b7yxj53j1rxw0000gq/C/clang/ModuleCache/OVVET4YQAF2U/_Builtin_stddef_max_align_t-E8YDO9UD74YG.pcm\'\n/private/tmp/swift-test-20220809-75938-1pkfhpt/swiftpmtest/Package.swift:4:8: error: failed to build module \'PackageDescription\' for importation due to the errors above; the textual interface may be broken by project issues or a compiler bug\nimport PackageDescription\n       ^", diagnosticFile: nil)
```
https://github.com/Homebrew/homebrew-core/runs/7748711279?check_suite_focus=true#step:10:6888

Seems unrelated to Python, but not certain. Any ideas, @Bo98?

**Comment by Bo98** [comment]
Probably a bad file in that machine's module cache. That's annoying. It shouldn't affect ephemeral at least.

I do pass a custom `-module-cache-path` in the tests, but we probably need it for the SwiftPM tests too. We could abuse this: https://github.com/apple/swift-package-manager/blob/release/5.6/Sources/PackageLoading/ManifestLoader.swift#L775

Before the SwiftPM test, try `ENV["SWIFTPM_MODULECACHE_OVERRIDE"] = module_cache`.

**Comment by carlocab** [comment]
I plan to rebase, update `resources`, and re-run this as soon as https://github.com/Homebrew/homebrew-core/pull/107645 is merged. The other fixes look like they'll take a bit longer.

**Comment by carlocab** [comment]
This is ready for review.

Some dependents will still fail, but all of them should have fixes in progress. See https://github.com/Homebrew/homebrew-core/pull/107517#issuecomment-1211000993.

I hope to be able to merge this once CI completes.

File: Formula/python@3.10.rb
Comment (MikeMcQuaid) [inline Formula/python@3.10.rb:61]: Does this mean whichever version is linked last will be `Current`?

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me. One small question.

File: Formula/python@3.10.rb
Comment (carlocab) [reply Formula/python@3.10.rb:61]: I contemplated that, but found it too confusing, so decided against it.

In this PR, only the newest Python in Homebrew/core (i.e. `python@3.10`) will have these symlinks.

I would like to find a way to do it so that the newest Python you have installed will have these symlinks, so that the behaviour is consistent and predictable, but im punting on that for now.

Comment (MikeMcQuaid) [reply Formula/python@3.10.rb:61]: > I contemplated that, but found it too confusing, so decided against it.

Agreed, I think it's better this way.

> In this PR, only the newest Python in Homebrew/core (i.e. `python@3.10`) will have these symlinks.

How do we ensure these get updated/migrated on a new Python version? Audit? Could do something like fail if the alias doesn't point to `python`.

Comment (carlocab) [reply Formula/python@3.10.rb:61]: > How do we ensure these get updated/migrated on a new Python version? Audit? Could do something like fail if the alias doesn't point to `python`.

Yup, we should add an audit for this. Will look at one soon.

Comment (MikeMcQuaid) [reply Formula/python@3.10.rb:61]: @carlocab Sounds great, doesn't need to block this PR 👍🏻

**Pull request closed without merging**
